inference-engine/src/mkldnn_plugin/nodes/extract_image_patches.cpp

```diff
@@ -1,248 +1,425 @@
+            case 2: uni_vpinsrw(xmm_src, xmm_src, op, 0x0); break;
+            case 1: uni_vpinsrb(xmm_src, xmm_src, op, 0x0); break;
+            default:
+                assert(!"unknown dtype size");
+        }
+    }
+    inline void store_scalar(const Xbyak::Address &op, Vmm vmm_arg) {
+        Xbyak::Xmm xmm_dst = Xmm(vmm_arg.getIdx());
+        switch (jpp.dtype_size) {
+            case 8: uni_vmovsd(op, vmm_arg); break;
+            case 4: uni_vmovss(op, vmm_arg); break;
+            case 2: uni_vpextrw(op, xmm_dst, 0x0); break;
+            case 1: uni_vpextrb(op, xmm_dst, 0x0); break;
+            default:
+                assert(!"unknown dtype size");
         }
     }
 
-    StatusCode execute(std::vector<Blob::Ptr>& inputs, std::vector<Blob::Ptr>& outputs, ResponseDesc *resp) noexcept override {
-        switch (inputs[0]->getTensorDesc().getPrecision().size()) {
-            case 1: {
-                process_data<PrecisionTrait<Precision::U8>::value_type>(inputs, outputs);
-                break;
-            }
-            case 2: {
-                process_data<PrecisionTrait<Precision::U16>::value_type>(inputs, outputs);
-                break;
-            }
-            case 4: {
-                process_data<PrecisionTrait<Precision::I32>::value_type>(inputs, outputs);
-                break;
-            }
-            case 8: {
-                process_data<PrecisionTrait<Precision::U64>::value_type>(inputs, outputs);
-                break;
-            }
-            default: {
-                if (resp) {
-                    std::string errorMsg = "ExtractImagePatches layer does not support precision '"
-                            + std::string(inputs[0]->getTensorDesc().getPrecision().name()) + "'";
-                    errorMsg.copy(resp->msg, sizeof(resp->msg) - 1);
-                }
-                return GENERAL_ERROR;
-            }
+    inline void pad_with_zeros(reg64_t &reg_num_pads_arg, reg64_t &reg_dst_arg) {
+        Xbyak::Label main, tail, exit;
+        L(main);
+        {
+            cmp(reg_num_pads_arg, jpp.block_size);
+            jl(tail);
+            uni_vmovups(ptr[reg_dst_arg], vmm_zero);
+            add(reg_dst_arg, jpp.dtype_size * jpp.block_size);
+            sub(reg_num_pads_arg, jpp.block_size);
+            jmp(main);
         }
+        L(tail);
+        {
+            cmp(reg_num_pads_arg, 0);
+            jle(exit);
+            store_scalar(ptr[reg_dst_arg], vmm_zero);
+            add(reg_dst_arg, jpp.dtype_size);
+            dec(reg_num_pads_arg);
+            jmp(tail);
+        }
+        L(exit);
+    }
 
-        return OK;
+    inline void read_src2vmm(const Xbyak::Xmm &xmm_arg, reg64_t &reg_src_arg) {
+        const int xmm_size = 16; // bytes
+        const int xmm_block_size = xmm_size / jpp.dtype_size;
+        for (int i = 0; i < xmm_block_size; i++) {
+            Xbyak::Address addr = ptr[reg_src_arg + i * jpp.SW * jpp.dtype_size];
```

Comment:
Have you tried gather instruction here?

Reply:
Thank you, I'll reimplement using gather instructions, when available.